### PR TITLE
VarHandles - Remove typeTable

### PR DESCRIPTION
**Pull request by bjornvar**
Previously, we kept an array of MethodTypes (`typeTable`) for each VarHandle, where each index in the array matched the `type` of the MethodHandle at the same index in the `handleTable`, except for the last argument. This was done so that we could keep extra arguments in the internal method, while exposing the expected public API. For example:
    `public int getAndSet(int value)` --> `private static int getAndSet(int value, VarHandle vh)`

In order to simplify the design, both in the VM and for future JIT optimizations, we have decided to remove the `typeTable`, and instead store a MethodType with the added VarHandle argument per call site.

The other changes in this PR are side-effects of removing the `typeTable` (plus some general VarHandle code cleanup).

### CODE REVIEW

**Comment by bjornvar** [comment]
@tajila Can you review this change?

**Comment by bjornvar** [comment]
@liqunl Can you review the JIT code?

**Comment by liqunl** [comment]
Can you fix the copyright failure? Otherwise the JIT change LGTM.

**Comment by bjornvar** [comment]
@DanHeidinga Can you review/merge this?

File: jcl/src/java.base/share/classes/java/lang/invoke/VarHandle.java [hunk 1]
```diff
@@ -317,7 +315,13 @@ int getModifiers() {
 	 * @return The {@link MethodType} corresponding to the provided {@link AccessMode}.
 	 */
 	public final MethodType accessModeType(AccessMode accessMode) {
-		return typeTable[accessMode.ordinal()];
+		MethodType internalType = handleTable[accessMode.ordinal()].type;
+		int numOfArguments = internalType.arguments.length;
+		
+		// Drop the internal VarHandle argument
+		MethodType modifiedType = internalType.dropParameterTypes(numOfArguments - 1, numOfArguments);
```
[the remark this answers is not in the document]
Comment (bjornvar) [reply jcl/src/java.base/share/classes/java/lang/invoke/VarHandle.java]: It was used once in our VarHandle code, but it was unnecessary, so I will remove it. A `grep` shows that it's used a fair bit by the OpenJDK VarHandle implementation, which doesn't affect ours. The search didn't show any other uses in the JDK.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java [hunk 2]
```diff
@@ -360,6 +360,20 @@ public static MethodType fromMethodDescriptorString(String methodDescriptor, Cla
...
+	/**
+	 * Internal helper method for generating a MethodType from a descriptor string,
+	 * and append an extra argument type.
+	 * 
+	 * VarHandle call sites use this method to generate the MethodType that matches
+	 * the VarHandle method implementation (extra VarHandle argument).
+	 */
+	private static final MethodType fromMethodDescriptorStringAppendArg(String methodDescriptor, ClassLoader loader, Class<?> appendArgumentType) {
+		MethodType result = fromMethodDescriptorString(methodDescriptor, loader);
+		result = result.appendParameterTypes(appendArgumentType);
```
[the remark this answers is not in the document]
Comment (bjornvar) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java]: No, I didn't consider it. I'll see if I can make that work.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java [hunk 3]
```diff
@@ -360,6 +360,20 @@ public static MethodType fromMethodDescriptorString(String methodDescriptor, Cla
 		return mt;
 	}
 	
+	/**
+	 * Internal helper method for generating a MethodType from a descriptor string,
+	 * and append an extra argument type.
+	 * 
+	 * VarHandle call sites use this method to generate the MethodType that matches
+	 * the VarHandle method implementation (extra VarHandle argument).
+	 */
```
[the remark this answers is not in the document]
Comment (bjornvar) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java]: I'll add it, although I would claim that it's not used by builder.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (bjornvar) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java]: Fixed

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (bjornvar) [reply jcl/src/java.base/share/classes/java/lang/invoke/MethodType.java]: Fixed

**Comment by DanHeidinga** [comment]
@andrewcraik Can you review the jit changes?

**Comment by DanHeidinga** [comment]
Jenkins test sanity

**Pull request merged**